5/README.md

```diff
@@ -22,46 +28,76 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
+This specification is intended to formally document the Tango Property model.
+Runtime representation of Property in conforming Tango implementations MUST follow this specification.
 
 ### Goals
 
- The specification of Property aims to define a standard structure and type to use in any implementation of Tango.
+A Property represents a configuration parameter in Tango.
 
-Additionally, it aims to:
-
-* Provide the initial information when one starts a Tango Device
-
-* Be usable as keeping global variable in the Tango control system
-
-* Provide meta data complementing any data in Tango
+There are 4 kinds of Tango Properties:
 
+ *  Class Property: a Property related to a Device Class (RFC-2).
+ *  Device Property: a Property related to a Device (RFC-2).
+ *  Attribute Property: a Property related to an Attribute (RFC-4).
+ *  Free Property: a Property defined for the entire Control System (RFC-1) i.e not related to a specific Class, Device or Attribute.
 
 ### Use Cases
 
-There are many use cases for the usage of a Property:
-
-* To connect a real network equipment a Property can define the target IP address
-
-* To change the representation of an information in order to be more user friendly a Property can define a new data format for the graphical user interface to convert the raw data.
+Here are some Tango Property use cases:
 
+* To connect a real network equipment a Property can define the target IP address or hostname
+* To change the representation of an information in order to be more user friendly a Property can define a new data format for the graphical user interface to convert the raw data
+* To configure some alarms on Attribute Value (See RFC-4)
+* To store the Attribute value of a Memorized Attribute (See RFC-4)
+* To configure the labels associated to a Tango:DevEnum attribute values
+* ...
 
 ## Specification
 
-A Tango Property is a strict definition of a pair of key/value
-* configuration are represented in the form of properties.
-* data are represented in the form of attributes.
-* actions are represented in the form of commands.
-
-Its model can be represented as a defined tree which each elements are from a defined types: Class, Device, Property, Attribute, Command following the rules below:
-
-
-A Tango Property is a strict definition of a pair of key/value
-* The Property SHALL have one key, called Property Name
-* The Property SHALL have one value
+A Tango Property is a strict definition of a pair of key/value:
+  * The Property SHALL have one key, called Property Name
+  * The Property SHALL have one value, which could be empty, called Property Value
+
+A Tango Property MAY be persisted in the Tango Database (See RFC-1 and RFC-6) or into a file if no Tango Database is used.
+
+Tango Device Properties and Tango Class Properties have a Type metadata.
+The following Types SHALL be supported by Device Properties and Class Properties (See RFC-9):
```

Comment:
> Actually Property Value are only string.

Property values are currently persisted as strings, in our **current** implementation of the Database server.
We could imagine having another Database Server implementation storing them as something different than a string.

In cppTango implementation, there is an abstraction class named DbDatum. There is also another type named DbData. which is a vector of DbDatum.
When we try to get a property value from the Database, we invoke some methods which will put the property value inside a DbDatum object at the end.

The DbDatum implementation provides some extraction operators to be able to extract the property value to many different types.

So, in theory, we could persist the properties in any format we want and keep the same interface on the C++ side (I don't know for the other languages).

In C++, the code generated by Pogo to get the properties is using the DbData/DbDatum classes and these extraction operators (operator>>) to convert the DbDatum to the property types defined by Pogo for these properties.

The main issue here, is to clearly define the scope of these specifications.
If we place ourselves from the device server programmer point of view, we should not say that the property values are strings. I think we don't need to put this restriction.
The tricky question is: from which point of view should we define these specifications?

> A Tango client is unable to know the type of Property.

Indeed. This is currently the case. It could probably be useful to have it in the future.
The Device server on the other side, knows the type of the property.

> Adding property from Jive does not require a type. db.get_device_property('...','...') does not reply with the expected type.

True for this part. But then the DbDatum extraction operators are used to extract the DbDatum to the expected property type. The property type is something known by the device server because the property type is something we define with POGO. It is part of the definition of the device.

> From I understand it's only implementation specific to convert property into a variable with an expected type.

Since the property type is something which is part of the device definition (it is something stored in the POGO xmi file), I thought it would be useful to precise that at least the property types currently supported by POGO SHALL be supported.

Reply:
I am agree that has to change in the future and this part will fit the purpose.
Yet today Property in the Tango protocol has no type (unfortunately ;-) . You can set a property with a string even if the Tango Device implementation expect a float.
And if a tango implementation use or not type for property it does not change anything for the rest of the Tango control system.
xmi file is only used by pogo i.e the python hl api does not use it.

Reply:
> I am agree that has to change in the future and this part will fit the purpose.
> Yet today Property in the Tango protocol has no type (unfortunately ;-) . 

But what is the definition of the Tango protocol? Aren't we describing the concepts here too?
I think the property type is part of the Tango concepts.

> You can set a property with a string even if the Tango Device implementation expect a float. 

True. But you can set it with a double too ;-). Still in the concept, a device property has a type.
This drives the behaviour of the device server. If you change the type of a device property and you don't change the value in the DB, the behaviour might change because the persisted property value might be interpreted differently. So the type is important.

> There is no mechanism to protect in the Tango world unless explicitly in the code of the Device (generated or not).

True. There is no strong restrictions here. Still I would not say that property values are string.

One of the original and main goals of a Device property is to be used in a Device server code to configure it. The type is important for the device server, at least in C++ and Java.

> And if a tango implementation use or not type for property it does not change anything for the rest of the Tango control system.

Still for the device server programmer, you should be able to store properties into variables having certain types. And this is one of the key features in some languages.
I still have difficulties to draw the lines of the scope of these specifications. :-(

> xmi file is used by pogo to generate developer code and create helper to convert easily to an expected type. It's not like an Attribute where to have to specify the type.

Tango was originally designed with Pogo in mind.
There are many Tango concepts and features which are there thanks to Pogo.
It is the same for the _mandatory in database_ properties for instance. It is the generated code from Pogo which is doing the check that the property value is stored in the DB.
But still I think the concept of _mandatory in database_ property is a feature of Tango, even though it does not necessary need to be implemented in a Tango kernel library. The combination of Pogo and kernel library provides this feature.
The xmi file defining a Tango device is something which can be used as a basis to regenerate code for another implementation of Tango.

The difficulty is to define where we draw the line for these specifications:
From a C++ device server programmer point of view, a persisted property value is a DbDatum, which can eventually be converted to a supported property type.
From the cppTango library point of view, the library gets the value of the persisted device properties for instance, via the result of DbGetDeviceProperty command executed on the TangoDatabase server, which is returning a DevVarStringArray.
The library will take care of converting this DevVarStringArray into a DbData.

This means that if we don't want to change the cppTango library, a compatible DatabaseDs device server should provide this DbGetDeviceProperty command which returns a DevVarStringArray.

On the other hand, if we consider only the source code compatibility on the device server side, we could imagine changing the way cppTango, JTango and pyTango retrieve the persisted properties. these libraries could talk to another service for instance, using a command different than DbGetDeviceProperty on the DatabaseDs, to retrieve device properties. It could even talk to different services depending on some parameters and still the interface on the device server side could stay the same. This new implementation should obviously come with some tools to be able to edit and visualize easily the persisted data.

So where do we draw the line?

Reply:
> > You can set a property with a string even if the Tango Device implementation expect a float.
> 
> True. But you can set it with a double too ;-). Still in the concept, a device property has a type.
> This drives the behaviour of the device server. If you change the type of a device property and you don't change the value in the DB, the behaviour might change because the persisted property value might be interpreted differently. So the type is important.

Correction: the property type is important from a Device server programmer perspective.

Reply:
@hardion, it is true that the sentence:
"Device Property and Class Property have a Type metadata." is a bit confusing because you currently have no way to know this type from a client perspective.
The type is not persisted in the DB for instance and the device wizard from jive for instance does not provide this type so I understand your comment and that this sentence is a bit misleading.
What would you suggest?

Reply:
@bourtemb 
I recognise that my comment about Tango Protocol was inappropriate.
My suggestion would be to keep the  "type" part in a separate branch v10 and open an issue with the tag "v10". For me this a valid request. Maybe some questions will be raised like how to handle the migration 9 ->10 (type mandatory or not, if yes what default type ).